[FSDP][optim_state_dict][2/N] Add _get_fqn_to_fsdp_param_info to map from original FQN to flat_param

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89981
* #89980
* #89900
* __->__ #89899
* #89898

**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.

fegin committed:
[FSDP][optim_state_dict][2/N] Add a helper to map from original FQN to flat_param

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add a helper to map from original FQN to flat_param"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add a helper to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add a helper to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add a helper to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -21,14 +23,26 @@
 
+@dataclass
+class FSDPParamInfo:
+    state: nn.Module
```
Is this called `state` because after composable support, the type is either `nn.Module` or the composable state type?

fegin replied:
Yes.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -84,7 +98,7 @@ class _OptimStateKey(NamedTuple):
 
     unflat_param_names: Tuple[str, ...]
-    is_flat_param: bool
+    is_fsdp_wrapped: bool
```
nit: Should we call this `is_fsdp_managed` (since no more wrapping in the composable path)?

For FSDP + Dynamo, we decided on using the word "managed" as well.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1220,6 +1238,9 @@ def _optim_state_dict(
         else _get_param_id_to_param(optim)
     )
+    fqn_to_fsdp_param_info = _get_fqn_to_fsdp_param_info(
+        model, set(fqn for v in param_to_fqns.values() for fqn in v)
```
Can you explain why there is the second `for fqn in v` and also generally why we need this `dedup_shared_fqns` argument to `_get_fqn_to_fsdp_param_info`? (See my comment below at the function as well.)

```
set(fqn for v in param_to_fqns.values() for fqn in v)
```

fegin replied:
For `use_orig_params == False` case, there may be more than 1 FQN for a param (which is a flat param) and we need to unflatten the list. Hence the second `for`.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1290,36 @@ def _optim_state_dict(
+
+
+def _get_fqn_to_fsdp_param_info(
+    model: nn.Module, dedup_shared_fqns: Set[str]
```
The name `dedup_shared_fqns` is a bit unclear to me. Is the deduplication just happening because this is a `set`, but this data structure mainly just represents the FQNs to add as keys to the returned `fqn_to_param_info`?

fegin replied:
Add a comment. But in general, we only need to track the first fqn of a shared parameter. Same case for `_get_param_to_fqns()` and `param_to_fqns`.

fegin replied:
After checking the code, I realized this protection is redundant. Please check the comment I added into the code.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1229,20 +1250,20 @@ def _optim_state_dict(
-            param = param_id_to_param[param_id]
-            fsdp_state = flat_param_to_fsdp_state[param]
+        if optim_state_key.is_fsdp_wrapped:
+            fqn = optim_state_key.unflat_param_names[0]
```
Conceptual: Why do we only take 0th index here?

IIUC, `unflat_param_names` is a `list` of the FQNs corresponding to 1 `FlatParameter`. Do we take the 0th index because `fqn_to_fsdp_param_info` is a surjective map, where we can key by any FQN corresponding to the `FlatParameter` to get the same `FSDPParamInfo`?

fegin replied:
If there are multiple `unflat_param_names` (not use_orig_params), they share the same FSDPParamInfo. So the first `unflat_param_name` is sufficient to fetch the FSDPParamInfo for both use or not use orig_params cases.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -21,14 +23,26 @@
+class FSDPParamInfo:
+    state: nn.Module
+    flat_param: FlatParameter
+    fqn_indices: Dict[str, int]
```
Could we add a comment explaining what these indices mean? (My understanding is that this is a mapping from FQN from `FlatParameter._fqns` to the corresponding index in `FlatParameter._fqns`.)

It also looks like this is not used for this PR (but probably for the next?).

fegin replied:
Change to `param_indices`.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1290,36 @@ def _optim_state_dict(
+        module_fn,
+        return_fn,
+        fqn_to_param_info,
+    )
```
nit: Could you add to the PR description that the helper you added is `_get_fqn_to_fsdp_param_info` just to make that explicit?

awgu reviewed (commented):
I made an initial pass with some conceptual questions.

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add _get_fqn_to_fsdp_param_info to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add _get_fqn_to_fsdp_param_info to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add _get_fqn_to_fsdp_param_info to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1289,39 @@ def _optim_state_dict(
+
+def _get_fqn_to_fsdp_param_info(model: nn.Module) -> Dict[str, FSDPParamInfo]:
+    """
+    Construct the maaping from a param's fqn to its corresponding FSDPParamInfo
```
nit:
```suggestion
    Construct the mapping from a param's fqn to its corresponding FSDPParamInfo
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1289,39 @@ def _optim_state_dict(
+def _get_fqn_to_fsdp_param_info(model: nn.Module) -> Dict[str, FSDPParamInfo]:
+    """
+    Construct the maaping from a param's fqn to its corresponding FSDPParamInfo
+    if the param is managed by FSDP. FlatParam only stores the first FQN of a
```
nit:
```suggestion
    if the param is managed by FSDP. FlatParamInfo only stores the first FQN of a
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1289,39 @@ def _optim_state_dict(
+    """
+    Construct the maaping from a param's fqn to its corresponding FSDPParamInfo
+    if the param is managed by FSDP. FlatParam only stores the first FQN of a
+    shared parameter. So the keys in the mapping are guranteed to map to unique
```
nit:
```suggestion
    shared parameter. So the keys in the mapping are guaranteed to map to unique
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1289,39 @@ def _optim_state_dict(
+    shared parameter. So the keys in the mapping are guranteed to map to unique
+    parameters.
+    """
+    def module_fn(module, prefix, fqn_to_param_info):
```
Should we add a comment saying we need to use `_apply_to_modules` to get the global FQN (since the saved FQNs are like local FQNs, not necessarily prefixed from the global root module)?

awgu reviewed (approved):
LGTM! I left some nits.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1269,3 +1289,39 @@ def _optim_state_dict(
+def _get_fqn_to_fsdp_param_info(model: nn.Module) -> Dict[str, FSDPParamInfo]:
+    """
+    Construct the maaping from a param's fqn to its corresponding FSDPParamInfo
+    if the param is managed by FSDP. FlatParam only stores the first FQN of a
```
nit: I got confused at first, but I think I understand:
```suggestion
    if the param is managed by FSDP. FlatParameter._fqns only stores the first FQN of a
```
(add backticks if you want, but maybe not to be consistent with rest of the comment)

fegin replied:
This is correct, thanks!

fegin committed:
Update on "[FSDP][optim_state_dict][2/N] Add _get_fqn_to_fsdp_param_info to map from original FQN to flat_param"


**Motivation:**
Add a helper to map from the FQN to the corresponding flat_param. The helper will directly get flat_param from fsdp_state and flat_handler as flat_param is not registered to the module if `use_orig_params` is True.


[ghstack-poisoned]

fegin commented:
@pytorchbot merge -f "The failing test is not related."
